cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -531,65 +532,117 @@ private MetadataEntry convertRow(Row row) {
    * @param searchQuery the search query, which could be of two forms: [key]:[value] or just [value] and can have '*'
    *                    at the end for a prefix search
    * @param types the {@link MetadataSearchTargetType} to restrict the search to, if empty all types are searched
+   * @param sortInfo the {@link SortInfo} to sort the results by
+   * @param offset index to start with in the search results. To return results from the beginning, pass {@code 0}.
+   *               Only applies when #sortInfo is not {@link SortInfo#DEFAULT}
+   * @param limit number of results to return, starting from #offset. To return all, pass {@link Integer#MAX_VALUE}.
+   *              Only applies when #sortInfo is not {@link SortInfo#DEFAULT}
+   * @param numCursors number of cursors to return in the response. A cursor identifies the first index of the
+   *                   next page for pagination purposes. Only applies when #sortInfo is not {@link SortInfo#DEFAULT}
+   * @param cursor the cursor that acts as the starting index for the requested page. This is only applicable when
+   *               #sortInfo is not {@link SortInfo#DEFAULT}. If offset is also specified, it is applied starting at
+   *               the cursor. If {@code null}, the first row is used as the cursor
+   * @return a {@link SearchResults} object containing a list of {@link MetadataEntry} containing each matching
+   *         {@link NamespacedEntityId} with its associated metadata. It also optionally contains a list of cursors
+   *         for subsequent queries to start with, if the specified #sortInfo is not {@link SortInfo#DEFAULT}.
    */
-  @VisibleForTesting
-  List<MetadataEntry> search(String namespaceId, String searchQuery, Set<MetadataSearchTargetType> types) {
-    return search(namespaceId, searchQuery, types, SortInfo.DEFAULT);
+  public SearchResults search(String namespaceId, String searchQuery, Set<MetadataSearchTargetType> types,
+                              SortInfo sortInfo, int offset, int limit, int numCursors,
+                              @Nullable String cursor) {
+    if (SortInfo.DEFAULT.equals(sortInfo)) {
```

Comment:
ok now I see. Lets make sure to validate that a cursor is not given if sorting by relevance in the handler and return a 400 if so.